sycl/source/detail/device_impl.cpp

```diff
@@ -101,7 +104,10 @@ cl_device_id device_impl::get() const {
 }
 
 platform device_impl::get_platform() const {
-  return createSyclObjFromImpl<platform>(MPlatform);
+  if (PlatformImplPtr Platform = MPlatform.lock())
```

Comment:
I think it's not very clear what should be done if a platform is released before a device bound to this platform.
Can we do vice versa - device has a shared_ptr to platform,  platform has a weak_ptr to device?

Reply:
I can try it.
  The platform has multiple owners "higher up", like the context and the GlobalHandler.  So it isn't affected if the device has a weak_ptr to it.   But I'm less certain for the devices. I think the platform might be the primary owner, so if it's a weak pointer, then the device might be released before the platform is done with it.   Also, the code has a clear chain of ownership now:  context->platform->device, it seems unwise to meddle with it.   

I would prefer that instead of returning just platform() in the else case, that we throw an exception - because no one should be asking for the platform after it has been released. However, that would be an ABI change, and I wanted to avoid anything controversial.

Reply:
I changed this PR to draft, and opened a new one here: https://github.com/intel/llvm/pull/2671